File: src/main/java/leetcode/offer/Offer21.java
package leetcode.offer;

/**
 * @author jingxinwu
 * @date 2021-11-21 6:05 下午
 */
public class Offer21 {

    public int[] exchange(int[] nums) {
        int i = 0, j = nums.length - 1, tmp;
        while (i < j) {
            //x&1 位运算 等价于 x%2 取余运算，即皆可用于判断数字奇偶性。
            while (i < j && (nums[i] & 1) == 1) {
                i++;
            }
            while (i < j && (nums[j] & 1) == 0) {
                j--;
            }
            tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
        }
        return nums;
    }

}

File: src/main/GO/leetcode/lists/offer/Offer21.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/28 23:40
 */

func exchange(nums []int) []int {
	left, right := 0, len(nums)-1
	for left < right {
		for left < right && nums[left]%2 == 1 {
			left++
		}
		for left < right && nums[right]%2 == 0 {
			right--
		}
		nums[left], nums[right] = nums[right], nums[left]
		left++
		right--
	}
	return nums
}
